src/oni2-keyboard-layout/stubs/keyboard-layout-linux.c

```diff
@@ -0,0 +1,222 @@
+} KeycodeMapEntry;
+
+#define USB_KEYMAP_DECLARATION static const KeycodeMapEntry keyCodeMap[] =
+#define USB_KEYMAP(usb, evdev, xkb, win, mac, code, id, sdlScancode, sdlKeycode) {xkb, code, sdlScancode, sdlKeycode}
+
+#include "keycode_converter_data.h"
+
+// In the atom JS lib, these are all attributes of the Manager class
+// Since we aren't using OOP, they can just be globals
+
+Display *xDisplay;
+XIC xInputContext;
+XIM xInputMethod;
+
+CAMLprim value oni2_KeyboardLayoutInit() {
+  CAMLparam0();
+
+  xDisplay = XOpenDisplay("");
+  if (xDisplay == NULL) {
+    RAISE_NULL;
+  }
+
+  xInputMethod = XOpenIM(xDisplay, 0, 0, 0);
+  if (xInputMethod == NULL) {
+    RAISE_NULL;
+  }
+  
+  XIMStyles *styles = NULL;
+  XGetIMValues(xInputMethod, XNQueryInputStyle, &styles, NULL);
+  if (styles == NULL) {
+    RAISE_NULL;
+  }
+
+  XIMStyle bestMatchStyle = 0;
+  for (int i = 0; i < styles->count_styles; i++) {
+    XIMStyle thisStyle = styles->supported_styles[i];
+    if (thisStyle == (XIMPreeditNothing | XIMStatusNothing)) {
+      bestMatchStyle = thisStyle;
+      break;
+    }
+  }
+
+  XFree(styles);
+  if (bestMatchStyle == 0) {
+    RAISE_NULL;
+  } 
+  
+  Window window;
+  int revertTo;
+  XGetInputFocus(xDisplay, &window, &revertTo);
+
+  if (window != BadRequest) {
+    xInputContext = XCreateIC(
+      xInputMethod,
+      XNInputStyle,
+      bestMatchStyle,
+      XNClientWindow,
+      window,
+      XNFocusWindow,
+      window,
+      NULL
+    );
+  }
+
+  CAMLreturn(Val_unit);
+}
+
+
+void oni2_priv_GetCurrentKeyboardLayout(char *layout) {
+  XkbRF_VarDefsRec vdr;
+  char *tmp;
+  if (XkbRF_GetNamesProp(xDisplay, &tmp, &vdr) && vdr.layout) {
+    XkbStateRec xkbState;
+    XkbGetState(xDisplay, XkbUseCoreKbd, &xkbState);
+
+    if (vdr.variant) {
+      sprintf(layout, "%s,%s[%d]", vdr.layout, vdr.variant, xkbState.group);
+    } else {
+      sprintf(layout, "%s[%d]", vdr.layout, xkbState.group);
+    }
+  } else {
+    layout[0] = '\0';
+  }
+}
+
+CAMLprim value oni2_KeyboardLayoutGetCurrentLayout() {
+  CAMLparam0();
+  CAMLlocal1(vLayout);
+
+  char layout[256];
+  oni2_priv_GetCurrentKeyboardLayout(layout);
+
+  vLayout = caml_copy_string(layout);
+
+  CAMLreturn(vLayout);
+}
+
+CAMLprim value oni2_KeyboardLayoutGetCurrentLanguage() {
+  CAMLparam0();
+  CAMLlocal1(vLanguage);
+
+  char language[256];
+  oni2_priv_GetCurrentKeyboardLayout(language);
+
+  vLanguage = caml_copy_string(language);
+
+  CAMLreturn(vLanguage);
+}
+
+void characterForNativeCode(
+  XIC xInputContext,
+  XKeyEvent *xKeyEvent, 
+  uint xkbKeycode, 
+  uint state,
+  char *buffer
+) {
+  xKeyEvent->keycode = xkbKeycode;
+  xKeyEvent->state = state;
+
+  if (xInputContext) {
+    int count = Xutf8LookupString(xInputContext, xKeyEvent, buffer, 2, NULL, NULL);
+    if (count > 0 && !iswcntrl(buffer[0])) {
+      buffer[count] = '\0';
+    } else {
+      buffer[0] = '\0';
+    }
+  } else {
+    char *cBuffer = (char *)buffer;
```

Comment:
I wonder why this is needed - can we just use `buffer` here? Or does it need to be copied?

Reply:
Oh my bad this is leftover from when I was using `wchar_t` for the buffers. I'll remove that line now.